packages/backend/server/src/plugins/copilot/session.ts

```diff
@@ -0,0 +1,216 @@
+  }
+
+  pop() {
+    this.state.messages.pop();
+  }
+
+  private takeMessages(): ChatMessage[] {
+    if (this.state.action) {
+      const messages = this.state.messages;
+      return messages.slice(messages.length - 1);
+    }
+    const ret = [];
+    const messages = [...this.state.messages];
+    messages.reverse();
+    let size = this.promptTokenSize;
+    for (const message of messages) {
+      const tokenSize = this.encoder.encode_ordinary(message.content).length;
+      if (size + tokenSize > this.maxTokenSize) {
+        break;
+      }
+      ret.push(message);
+      size += tokenSize;
+    }
+    ret.reverse();
+    return ret;
+  }
+
+  finish(): ChatMessage[] {
+    const messages = this.takeMessages();
+    return [...this.state.prompt, ...messages];
+  }
+
+  async save() {
+    await this.dispose?.(this.state);
+  }
+
+  async [Symbol.asyncDispose]() {
+    this.encoder.free();
+    await this.save?.();
+  }
+}
+
+@Injectable()
+export class ChatSessionService {
+  private readonly logger = new Logger(ChatSessionService.name);
+  constructor(
+    private readonly db: PrismaClient,
+    private readonly cache: SessionCache,
+    private readonly prompt: PromptService
+  ) {}
+
+  private async setSession(state: ChatSessionState): Promise<void> {
+    await this.db.aiSession.upsert({
+      where: {
+        id: state.sessionId,
+      },
+      update: {
+        messages: {
+          create: state.messages.map((m, idx) => ({ idx, ...m })),
+        },
+      },
+      create: {
+        id: state.sessionId,
+        action: state.action,
+        model: state.model,
+        messages: { create: state.messages },
+        // connect
+        user: { connect: { id: state.userId } },
+        workspace: { connect: { id: state.workspaceId } },
+        doc: {
+          connect: {
+            id_workspaceId: { id: state.docId, workspaceId: state.workspaceId },
+          },
+        },
+        promptName: state.promptName,
+      },
+    });
+  }
+
+  private async getSession(
+    sessionId: string
+  ): Promise<ChatSessionState | undefined> {
+    return await this.db.aiSession
+      .findUnique({
+        where: { id: sessionId },
+      })
+      .then(async session => {
+        if (!session) return;
+        const messages = ChatMessageSchema.array().safeParse(session.messages);
+
+        return {
+          sessionId: session.id,
+          userId: session.userId,
+          workspaceId: session.workspaceId,
+          docId: session.docId,
+          promptName: session.promptName,
+          action: session.action,
+          model: session.model as TiktokenModel,
+          prompt: await this.prompt.get(session.promptName),
+          messages: messages.success ? messages.data : [],
+        };
+      });
+  }
+
+  private async setState(state: ChatSessionState): Promise<void> {
+    await Promise.all([
+      this.cache.set(`${CHAT_SESSION_KEY}:${state.sessionId}`, state, {
```

Comment:
large key, here is context length, will significantly increase redis latency

Reply:
removed, will optimize query based on splited message table